src/Internal/QueryM/Kupo.purs

```diff
@@ -156,19 +165,25 @@ isTxConfirmedAff config (TransactionHash txHash) = runExceptT do
 
 getTxMetadata
   :: TransactionHash
-  -> QueryM (Either ClientError (Maybe GeneralTransactionMetadata))
+  -> QueryM (Either GetTxMetadataError GeneralTransactionMetadata)
 getTxMetadata txHash = runExceptT do
-  ExceptT (isTxConfirmed txHash) >>= case _ of
-    Nothing -> pure Nothing
+  ExceptT (lmap GetTxMetadataClientError <$> isTxConfirmed txHash) >>= case _ of
+    Nothing -> throwError GetTxMetadataTxNotFoundError
     Just slot -> do
       let
         endpoint = "/metadata/" <> BigNum.toString (unwrap slot)
           <> "?transaction_id="
           <> byteArrayToHex (unwrap txHash)
-      metadata <- ExceptT $ handleAffjaxResponse <$> kupoGetRequest
-        endpoint
-      -- `KupoMetadata` will fail parsing if there is more than one response
-      pure $ unwrapKupoMetadata metadata
+      kupoMetadata <- ExceptT $
+        lmap GetTxMetadataClientError <<< handleAffjaxResponse <$>
+          kupoGetRequest
+            endpoint
+      case unwrapKupoMetadata kupoMetadata of
+        Nothing -> throwError GetTxMetadataMetadataEmptyOrMissingError
+        Just metadata
+          | Map.isEmpty (unwrap metadata) -> throwError
```

Comment:
Why we need to handle empty Map case?

Reply:
To make the API the same for blockfrost and kupo.

Blockfrost always gives a response for metadata, regardless of whether it actually had the optional metadata field attached. Kupo on the other hand returns exactly what the transaction contained,

The alternative could be to always make the kupo version return an empty metadata on the case where there was none, but I think if users are requesting metadata from a transaction, they probably expect some metadata.

Reply:
Not sure if I explained that well, maybe this makes it clearer:

No metadata:
blockfrost: []
kupo: []

Empty metadata:
blockfrost: []
kupo: [ ... ]

With metadata:
blockfrost: [ ... ]
kupo: [ ... ]